### PR TITLE
FreeBSD ready

### PR DESCRIPTION
**Pull request by ohmtal**
Compiles on FreeBSD but libs:  pcre  and convexDecomp are altered to work.  

Fixed: I have a problem with the posix CPUInfo, because std::ptr_fun is obsolete. 

Added posix fix for onExit. (https://torque3d.org/forums/topic/7261-fix-for-onexit-not-called-on-linuxbsd-posix/)

### CODE REVIEW

**Comment by Ragora** [comment]
I feel like there's probably a better way to do the trims I had in there (maybe not, though):

https://stackoverflow.com/questions/216823/how-to-trim-an-stdstring

```c++
// trim from start (in place)
static inline void ltrim(std::string &s) {
    s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](unsigned char ch) {
        return !std::isspace(ch);
    }));
}

// trim from end (in place)
static inline void rtrim(std::string &s) {
    s.erase(std::find_if(s.rbegin(), s.rend(), [](unsigned char ch) {
        return !std::isspace(ch);
    }).base(), s.end());
}
```

Torque might have a usable function as well, but I dunno what it would be offhand.

**Comment by ohmtal** [comment]
> I feel like there's probably a better way to do the trims I had in there (maybe not, though):
> 
> https://stackoverflow.com/questions/216823/how-to-trim-an-stdstring
> 
> ```c++
> // trim from start (in place)
> static inline void ltrim(std::string &s) {
>     s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](unsigned char ch) {
>         return !std::isspace(ch);
>     }));
> }
> 
> // trim from end (in place)
> static inline void rtrim(std::string &s) {
>     s.erase(std::find_if(s.rbegin(), s.rend(), [](unsigned char ch) {
>         return !std::isspace(ch);
>     }).base(), s.end());
> }
> ```
> 
> Torque might have a usable function as well, but I dunno what it would be offhand.

I've added your changes.

File: Engine/lib/convexDecomp/NvSimpleTypes.h [hunk 1]
```diff
@@ -87,7 +92,7 @@ NvSimpleTypes.h : Defines basic data types for integers and floats.
 	typedef float				NxF32;
 	typedef double				NxF64;
 		
-#elif defined(LINUX)
+#elif LINUX
```
Comment (dottools) [inline Engine/lib/convexDecomp/NvSimpleTypes.h:95]: What is going on here with removing `defined()` evaluations?

File: Engine/lib/convexDecomp/NvThreadConfig.cpp [hunk 2]
```diff
@@ -129,26 +133,22 @@ void   tc_sleep(NxU32 ms)
...
-#elif defined( _WIN64 )
-   YieldProcessor( );
-#elif defined( __APPLE__ )
-   pthread_yield_np();
-#elif defined(__linux__)
-   #if defined(_POSIX_PRIORITY_SCHEDULING)
-      sched_yield();
+   #if defined(__linux__) || defined( __APPLE__ ) ||  defined( __FreeBSD__)
+
+      asm ( "pause" );
```
Comment (dottools) [inline Engine/lib/convexDecomp/NvThreadConfig.cpp:138]: Pretty sure we've already previously had discussions about using inlined x86 assembly instruction, vs using the preferred way of involving the OS scheduler in yielding the task and for cross architecture compatibility (`pause` doesn't exist for ARM, etc..).

File: Engine/source/platformPOSIX/POSIXCPUInfo.cpp [hunk 3]
```diff
@@ -20,7 +20,8 @@
 // IN THE SOFTWARE.
 //-----------------------------------------------------------------------------
 
-#ifndef __APPLE__
+//XXTH orig #ifndef __APPLE__
+#if not defined (__APPLE__) //and not defined (__FreeBSD__)
```
Comment (dottools) [inline Engine/source/platformPOSIX/POSIXCPUInfo.cpp:24]: `#if not defined (__APPLE__)` is not a valid/standard preprocessor expression.

I believe you meant
`#if !defined (__APPLE__)`

**Comment by ohmtal** [comment]
Thanks for the review dottools. I think it need some cleanup. Will close it for now.

**Pull request closed without merging**